src/search/landmarks/landmark_status_manager.cc

```diff
@@ -106,69 +117,80 @@ bool LandmarkStatusManager::update_reached_lms(const GlobalState &parent_global_
+        int id = node->get_id();
 
-    // mark reached and find needed again landmarks
-    for (auto &node : nodes) {
-        if (node->status == lm_reached) {
-            if (!node->is_true_in_state(global_state)) {
-                if (node->is_goal()) {
-                    node->status = lm_needed_again;
-                } else {
-                    if (check_lost_landmark_children_needed_again(*node)) {
-                        node->status = lm_needed_again;
-                    }
-                }
-            }
-        }
+        /*
+          This dead-end detection works for the following case:
+          X is a goal, it is true in the initial state, and has no achievers.
+          Some action A has X as a delete effect. Then using this,
+          we can detect that applying A leads to a dead-end.
 
-        // This dead-end detection works for the following case:
-        // X is a goal, it is true in the initial state, and has no achievers.
-        // Some action A has X as a delete effect. Then using this,
-        // we can detect that applying A leads to a dead-end.
-        //
-        // Note: this only tests for reachability of the landmark from the initial state.
-        // A (possibly) more effective option would be to test reachability of the landmark
-        // from the current state.
+          Note: this only tests for reachability of the landmark from the initial state.
+          A (possibly) more effective option would be to test reachability of the landmark
+          from the current state.
+        */
 
         if (!node->is_derived) {
-            if ((node->status == lm_not_reached) &&
+            if ((lm_status[id] == lm_not_reached) &&
                 node->first_achievers.empty()) {
-                dead_end_found = true;
+                return true;
             }
-            if ((node->status == lm_needed_again) &&
+            if ((lm_status[id] == lm_needed_again) &&
                 node->possible_achievers.empty()) {
-                dead_end_found = true;
+                return true;
             }
         }
     }
-
-    return dead_end_found;
+    return false;
 }
 
+bool LandmarkStatusManager::landmark_needed_again(
+    int id, const GlobalState &state) {
 
-bool LandmarkStatusManager::check_lost_landmark_children_needed_again(const LandmarkNode &node) const {
-    for (const auto &child : node.children) {
-        LandmarkNode *child_node = child.first;
-        if (child.second >= EdgeType::greedy_necessary && child_node->status == lm_not_reached)
-            return true;
+    LandmarkNode *node = lm_graph.get_lm_for_index(id);
+    if (node->is_true_in_state(state)) {
+        return false;
+    } else if (node->is_goal()) {
+        return true;
+    } else {
+        /* For all A ->_gn B, if B is not reached and A currently not
+           true, since A is a necessary precondition for actions
+           achieving B for the first time, it must become true again. */
```

Comment:
Is this new text or did it exist before? Just asking out of curiosity.

Note that the style for long comments is
/*
  text
  text
*/

i.e., slightly differently than it is here.

Reply:
The style guide sais it's fine if "short comments (2-3 lines)" are commented in this fashion, but I don't have a strong opinion on this, so I will change it.

Reply:
Ok, me neither, I just thought the limit was 2 lines. Never mind then.

Reply:
This text is new.

Reply:
I think these days, I always use the style that Silvan cites also if it's only two lines, and I think the code is moving towards this convention in general. I think the exception in the style guide is there primarily for people that felt uncomfortable about the extra lines back when we first discussed style, but over time most people have become comfortable with this way of commenting. This is a common thing. When we first debate such things, opinions are often very strong and we debate heavily over what should or shouldn't be allowed, but over time we often fall into patterns that make some of the old discussions redundand. So I have a preference for the version with start/end on separate lines, but I don't feel strongly about it.